File: CSCI 3180/2018/HW_2/task4_python/map.py
# CSCI3180 Principles of Programming Languages
#
# --- Declaration ---
#
# I declare that the assignment here submitted is original except for source
# material explicitly acknowledged. I also acknowledge that I am aware of
# University policy and regulations on honesty in academic work, and of the
# disciplinary guidelines and procedures applicable to breaches of such policy
# and regulations, as contained in the website
# http://www.cuhk.edu.hk/policy/academichonesty/
#
# Assignment 2
# Name : Wong Sin Yi
# Student ID : 1155110677
# Email Addr : 1155110677@link.cuhk.edu.hk

import random

from land import land
from pos import pos
from warrior import warrior
from elf import elf
from monster import monster
from potion import potion

class map:


    def __init__(self):
        self.__D = 10
        self.__teleportable_obj = []
        self.__e = random.randint(0,1) + 2
        self.__m = random.randint(0,1) + 2
        self.__p = random.randint(0,1) + 2
        self.__w = random.randint(0,1) + 2
        self.__total_num = self.__m + self.__e + self.__w + self.__p
        self.__num_of_alive_monsters = self.__m
        self.__num_of_alive_warriors = self.__w

    def initialize_all(self):  
        print('Welcome to Kafustrok. Light blesses you. ')
        self.__lands = []
        for i in range (0, self.__D):
            new = []
            for j in range (0, self.__D):
                new.append(land())
            self.__lands.append(new)

        for i in range (0, self.__total_num):
            pos = self.unoccupied_position
            if i < self.__m:
                self.__lands[pos.x][pos.y].occupied_obj = monster(pos.x, pos.y, i, self)
            elif i < self.__m + self.__e:
                self.__lands[pos.x][pos.y].occupied_obj = elf(pos.x, pos.y, i-self.__m, self)
            elif i < self.__m + self.__e + self.__p:
                self.__lands[pos.x][pos.y].occupied_obj = potion(pos.x, pos.y, i-self.__m - self.__e, self)
                self.__teleportable_obj.append(self.__lands[pos.x][pos.y].occupied_obj)
            else:
                self.__lands[pos.x][pos.y].occupied_obj = warrior(pos.x, pos.y, i-self.__m-self.__e -self.__p, self)
                self.__teleportable_obj.append(self.__lands[pos.x][pos.y].occupied_obj)

    def teleport_all(self):
        for obj in self.__teleportable_obj:
            if type(obj) is warrior:
                self.print_broad()
                obj.teleport()
                if self.num_of_alive_monsters == 0:
                    return
        for obj in self.__teleportable_obj:
            if type(obj) is potion:
                obj.teleport()

    def coming(self, posx, posy, warrior):
        return self.__lands[posx][posy].coming(warrior)

    def set_land(self,_pos, occupied_obj):
        self.__lands[_pos.x][_pos.y].occupied_obj = occupied_obj

    def delete_teleportable_obj(self, obj):
        index = self.__teleportable_obj.index(obj)
        self.__teleportable_obj[index] = None

    @property
    def unoccupied_position(self):
        randx = random.randint(0, self.__D - 1 )
        randy = random.randint(0, self.__D - 1 )
        while self.__lands[randx][randy].occupied_obj != None:
            randx = random.randint(0, self.__D - 1 )
            randy = random.randint(0, self.__D - 1 )
        return pos(randx, randy)

    def print_broad(self):
        printObject = []
        for i in range (0, self.__D):
            new = []
            printObject.append(new)

        for i in range (0, self.__D):
            for j in range (0, self.__D):
                occupant_name = self.__lands[i][j].occupant_name
                if occupant_name is None:
                    occupant_name = "  "
                printObject[i].append(occupant_name)

        print(" ", end = '')
        for i in range (0, self.__D):
            print("| "+ str(i) + "  ", end = '')
        print("|")

        for i in range (0, int(self.__D * 5.5)):
            print("-", end = '')
        print("")

        for row in range (0, self.__D):
            print(row, end = '')
            for col in range (0, self.__D):
                print("| "+ printObject[row][col] + " ", end = '')
            print("|")
            for i in range (0, int(self.__D * 5.5)):
                print("-", end = '')
            print("")

    def decrease_num_of_alive_monsters(self):
        self.__num_of_alive_monsters -= 1

    def decrease_num_of_warriors(self):
        self.__num_of_alive_warriors -= 1

    @property
    def num_of_alive_monsters(self):
        return self.__num_of_alive_monsters

    @num_of_alive_monsters.setter
    def num_of_alive_monsters(self , num_of_alive_monsters):
        self.__num_of_alive_monsters = num_of_alive_monsters

    @property
    def num_of_alive_warriors(self):
        return self.__num_of_alive_warriors

    @num_of_alive_warriors.setter
    def num_of_alive_warriors(self, num_of_alive_warriors):
        self.__num_of_alive_warriors = num_of_alive_warriors

File: CSCI 3180/2018/HW_2/task4_java/Map.java
/*
 * CSCI3180 Principles of Programming Languages
 *
 * --- Declaration ---
 *
 * I declare that the assignment here submitted is original except for source
 * material explicitly acknowledged. I also acknowledge that I am aware of
 * University policy and regulations on honesty in academic work, and of the
 * disciplinary guidelines and procedures applicable to breaches of such policy
 * and regulations, as contained in the website
 * http://www.cuhk.edu.hk/policy/academichonesty/
 *
 * Assignment 2
 * Name : Wong Sin Yi
 * Student ID : 1155110677
 * Email Addr : 1155110677@link.cuhk.edu.hk
 */

import java.util.ArrayList;

public class Map {
	
	private ArrayList<Object> teleportable_obj = new ArrayList<Object>();
	private Land[][] lands;
	private final int D = 10; //dimension of board
	private int e;// Number of elves
	private int m;//Number of monsters
	private int w;//Number of warriors
	private int p;//Number of Potion
	private int totalNum = 0;
	private int numOfAliveMonsters;
	private int numOfAliveWarriors;
	
	public Map() {
		lands = new Land[D][D];
		e = TheJourney.rand.nextInt(2) + 2;
		m = TheJourney.rand.nextInt(2) + 2;
		w = TheJourney.rand.nextInt(2) + 2;
		p = TheJourney.rand.nextInt(2) + 2;
		totalNum = m + e + w + p;
		this.numOfAliveMonsters = m;
		this.numOfAliveWarriors = w;
	}
	
	public void initializeAll() {
		System.out.println("Welcome to Kafustrok. Light blesses you. ");

		for (int i = 0; i < D; i++)
			for (int j = 0; j < D; j++)
				lands[i][j] = new Land();
		for (int i = 0; i < totalNum; i++) {
			Pos pos = getUnOccupiedPosition();
			if (i < m)
				this.lands[pos.getX()][pos.getY()].setOccupied_obj(new Monster(pos.getX(), pos.getY(), i, this));
			else if (i < m+e)
				this.lands[pos.getX()][pos.getY()].setOccupied_obj(new Elf(pos.getX(), pos.getY(), i-m, this));
			else if (i < m+e+p){
				this.lands[pos.getX()][pos.getY()].setOccupied_obj(new Potion(pos.getX(), pos.getY(), i-m-e, this));
                teleportable_obj.add(this.lands[pos.getX()][pos.getY()].getOccupied_obj());
			}else {
				this.lands[pos.getX()][pos.getY()].setOccupied_obj(new Warrior(pos.getX(), pos.getY(), i-m-e-p, this));
				teleportable_obj.add(this.lands[pos.getX()][pos.getY()].getOccupied_obj());
			}
		}
	}
	
	public void teleportAll() {
		for (Object obj : teleportable_obj) {
		    if (obj instanceof Warrior) {
				this.printBoard();
				((Warrior) obj).teleport();
                if(this.numOfAliveMonsters == 0)
					return;
			}
		}

		for (Object obj : teleportable_obj) {
			if (obj instanceof Potion){
                ((Potion)obj).teleport();
			}
		}
	}
	
	public boolean coming(int posx, int posy, Warrior warrior) {
		return this.lands[posx][posy].coming(warrior);
	}
	
	public void setLand(Pos _pos, Object occupied_obj) {
		this.lands[_pos.getX()][_pos.getY()].setOccupied_obj(occupied_obj);
	}
	
	public void deleteTeleportableObj(Object obj) {
		int index = this.teleportable_obj.indexOf(obj);
		this.teleportable_obj.set(index, null);
	}
	/**
	 * Make sure all position is unique
	 * @return an unique position
	 */
	public Pos getUnOccupiedPosition() {
    	int randx = TheJourney.rand.nextInt(D);
		int randy = TheJourney.rand.nextInt(D);
    	while (this.lands[randx][randy].getOccupied_obj() != null) {    
        	randx = TheJourney.rand.nextInt(D);
    		randy = TheJourney.rand.nextInt(D);
		}
		return new Pos(randx,randy);
    }
    
    
	public void printBoard() {
		String printObject[][] = new String[D][D];

		// initialize printObject
		for (int i = 0; i < D; i++) {
			for (int j = 0; j < D; j++) {
				String occupantName = lands[i][j].getOccupantName();
				if (occupantName == null)
					occupantName = "  ";
				printObject[i][j] = occupantName;
			}
		}

		// printing
		System.out.print(" ");
		for (int i = 0; i < D; i++)
			System.out.print(String.format("| %d  ", i));

		System.out.println("|");

		for (int i = 0; i < D * 5.5; i++)
			System.out.print("-");
		System.out.println("");

		for (int row = 0; row < D; row++) {
			System.out.print(row);
			for (int col = 0; col < D; col++)
				System.out.print(String.format("| %s ",
				printObject[row][col]));
			System.out.println("|");
			for (int i = 0; i < D * 5.5; i++)
				System.out.print("-");
			System.out.println("");
		}

	}
    
	public void decreaseNumOfAliveMonsters() {
		this.numOfAliveMonsters--;
	}
	public void decreaseNumOfWarriors() {
		this.numOfAliveWarriors--;
	}
	
	public int getNumOfAliveMonsters() {
		return numOfAliveMonsters;
	}

	public void setNumOfAliveMonsters(int numOfAliveMonsters) {
		this.numOfAliveMonsters = numOfAliveMonsters;
	}

	public int getNumOfAliveWarriors() {
		return numOfAliveWarriors;
	}

	public void setNumOfAliveWarriors(int numOfAliveWarriors) {
		this.numOfAliveWarriors = numOfAliveWarriors;
	}
}
